libraries/AC_PID/AC_P_1D.cpp

```diff
@@ -0,0 +1,171 @@
+/// @file	AC_P_1D.cpp
+/// @brief	Generic P algorithm
+
+#include <AP_Math/AP_Math.h>
+#include "AC_P_1D.h"
+
+const AP_Param::GroupInfo AC_P_1D::var_info[] = {
+    // @Param: P
+    // @DisplayName: P Proportional Gain
+    // @Description: P Gain which produces an output value that is proportional to the current error value
+    AP_GROUPINFO("P",    0, AC_P_1D, _kp, 0),
+    AP_GROUPEND
+};
+
+// Constructor
+AC_P_1D::AC_P_1D(float initial_p, float dt) :
+    _dt(dt)
+{
+    // load parameter values from eeprom
+    AP_Param::setup_object_defaults(this, var_info);
+
+    _kp = initial_p;
+    _error = 0.0f;
+    _lim_err_neg = 0.0f;     // error value to enable filtering
+    _lim_err_pos = 0.0f;     // error value to enable filtering
+    _lim_out_neg = 0.0f;     // error value to enable filtering
+    _lim_out_pos = 0.0f;     // error value to enable filtering
+    _lim_D_Out = 10.0f;     // maximum first differential of output
+}
+
+// set_dt - set time step in seconds
+void AC_P_1D::set_dt(float dt)
+{
+    // set dt and calculate the input filter alpha
+    _dt = dt;
+}
+
+// set_dt - set time step in seconds
+void AC_P_1D::set_limits_error(float lim_err_neg, float lim_err_pos, float lim_out_neg, float lim_out_pos, float lim_D_Out, float lim_D2_Out)
+{
+    _lim_err_neg = -inv_sqrt_controller(lim_err_neg, _kp, _lim_D_Out);
+    _lim_err_pos = inv_sqrt_controller(lim_err_pos, _kp, _lim_D_Out);
+    _lim_out_neg = lim_out_neg;     // error value to enable filtering
+    _lim_out_pos = lim_out_pos;     // error value to enable filtering
+    _lim_D_Out = lim_D_Out;
+    if(is_positive(lim_D2_Out)) {
+        // limit the first derivative so as not to exceed the second derivative
+        _lim_D_Out = MIN(_lim_D_Out, lim_D2_Out / _kp);
+    }
+}
+
+// update_all - set target and measured inputs to PID controller and calculate outputs
+// target and measurement are filtered
+// if measurement is further than error_min or error_max (see set_limits_error method)
+//   the target is moved closer to the measurement and limit_min or limit_max will be set true
+float AC_P_1D::update_all(float &target, float measurement, bool &limit_min, bool &limit_max)
```

Comment:
Not really sure we need the `_all` ?

Similarly on the other pid objects

Reply:
This comes from a past in which the caller was responsible for updating the P, I and D separately.  This is no longer required but I think I'll leave the name as-is to avoid too much renaming and to keep the classes consistent.